### ISSUE TITLE
Stop using Common Names for matching

### ISSUE THREAD

**Issue by wrygiel** [erasmus-without-paper/ewp-specs-architecture#2]
In https://github.com/erasmus-without-paper/ewp-specs-api-echo/issues/2 @andydowling-ie wrote:

> We need to be careful when performing certificate matching. Considering the Registry is storing actual copies of the X.509 certs in the catalogue.xml, then its better to match based on cert fingerprint, rather than Common Name. (Although Common Name can be used to index the certs when searching the Registry for a match).

@wrygiel:

> We have been considering such scenario and concluded that it will cause problems when certificates on the server are upgraded (the server could be temporarily cut off from the network, if the manifest was not handled properly). We would need to analyze if the risk of a bad implementation outweighs the benefits of continuous availability. Should I start a separate issue for that?

@andydowling-ie:

> If you compare using just Common Name, your security is dependent on all CAs in your trusted root store not issuing any maliciously or incorrectly-named certs. It's probably a fair assumption, but it you're paranoid, then comparing the cert fingerprint (or cert bytes) is stronger. Happy to discuss via separate thread.

**Comment by wrygiel** [comment]
> Considering the Registry is storing actual copies of the X.509 certs in the catalogue.xml

Currently, the registry is storing **client** certificates **only**, and the recommended option of providing them is the `<common-name>` element. The `<certificate>` element was added primarily for self-signed certificates, and it is discouraged in production environment.

> your security is dependent on all CAs in your trusted root store not issuing any maliciously or incorrectly-named certs. It's probably a fair assumption, but it you're paranoid, then comparing the cert fingerprint (or cert bytes) is stronger.

We also deemed it a fair assumption. Do you really think we really should consider the scenario of rogue CAs? Somehow I fear that introducing server certificates in the manifest file could introduce more problems than it would solve...

**Comment by andydowling-ie** [comment]
I agree that there should be a distinction between how client and server certificates are treated. I'll try to articulate....

Right now, we have a registry that contains a list of all HEIs (i.e. the `catalogue.xml` contains the `id` attribute inside the `hei` element). This is a necessary part of the architecture, as it gives clients an authoritative list of participating HEIs and their server names. These can be matched by a connecting client and the client can check for a valid SSL server certificate with a matching hostname. I'm fine with this, as to mount an effective attack, and attacker would likely need to:

a. Poison a client's route to DNS to redirect that client to their "bad" server which they control (and have the private key for)
b. Use a malicious certificate to masquerade as a legitimate EWP server to the client

...so that the client can't tell the difference. That's a fairly complex attack. If it was DNS poisoning - that's a critical issue that would affect the entire internet, so there are likely to be bigger targets out there than EWP clients. My view is that storing server certificates in the registry might be overkill in this case.

However, the situation with client certificates is quite different. I'm assuming here that EWP hosts are not protected by firewalls and that they are wide open to the internet. That means that the only layer of authentication is that provided by the client SSL certificate. This makes perfect sense as we don't want each EWP host to have to store authentication credentials (username/password, etc), or otherwise deal with synchronisation issues across EWP hosts, etc.

This means that any internet client can try and connect using a certificate + key that it possesses. The decision to allow/deny access is based purely on the client certificate matching algorithm. It we allow any client certificate that is trusted by our server's trust store, and only check the client cert's Subject DN, then we're exposed if any of the CAs in our trust store have been compromised and the client is connecting using a fake cert.

This can happen in a variety of ways:
1. CA certificate has been brute-force attacked. This happened back in 2008 and caused CAs to deprecate use of MD5 for signing certs: http://www.zdnet.com/article/ssl-broken-hackers-create-rogue-ca-certificate-using-md5-collisions/
2. CA certificate has been compromised and somebody is issuing rogue certificates. It happened in India in 2014: http://www.pcworld.com/article/2453343/microsoft-revokes-trust-in-certificate-authority-operated-by-the-indian-government.html

Either way, the point is that by checking Subject DNs only, EWPs are effectively delegating their trust to a whole bunch of CAs, not all of which might be 100%. So, it's a risk for sure, it's just a question of how far we want to go to address it.

You might consider some of the following options:

**Option A**. Stipulating that client certs are issued by CAs that meet certain criteria (in terms of algorithms, cert issuing policies, etc.). Server implementations of EWP would need to cater for this.

**Option B**. Include all supported client certificates/fingerprints in the catalogue as you have done, and that they are properly matched (i.e. fingerprint match at least) in all cases by EWP server implementations.

By including client certs in the registry, you put the power back in the hands of the EWP servers to enforce security. However it does come at a cost - those client certs will need to be updated when they expire, are revoked, etc.

Hope that clarifies.

**Comment by wrygiel** [comment]
> By including client certs in the registry, you put the power back in the hands of the EWP servers to enforce security. However it does come at a cost - those client certs will need to be updated when they expire, are revoked, etc.

Thank you, that's a very comprehensive answer. It seems it all comes down to the security vs. usability debate once again! I will think about it a little more, and possibly come up with some more options.

**Comment by wrygiel** [comment]
Let's think about **option B** for a moment and try to eliminate the drawbacks it would cause. I believe @vgiralt and @Awerin would be interested in this too.

> those client certs will need to be updated when they expire, are revoked, etc.

I came up with something like this. Please verify:

**Option B+**. I believe we could make use of the fact that **public key fingerprints** (as opposed to certificate fingerprints) [usually](http://security.stackexchange.com/questions/27810/should-i-change-the-private-key-when-renewing-a-certificate) don't change after the certificate is renewed. We could include such fingerprints in the Registry response, instead of including whole certificates. This way, if the admin forgets to update the certificate in his manifest after its renawal - it still won't break anything (because the registry will be serving public key fingerprints only).

In other words we would replace this:

``` xml
<certificate fingerprint="831a78e44453bbe554d942a0c33a6e59f248c9af">
    MIIB9TCCAV4CCQDzcL/uks2ttjANBgkqhkiG9w0BAQUFADA/MQswCQYDVQQGEwJOTzETMBEGA1UE
    CAwKU29tZS1TdGF0ZTENMAsGA1UEBwwET3NsbzEMMAoGA1UECgwDVWlPMB4XDTE1MDIwNTEzNDE1
    MloXDTE4MDUyMDEzNDE1MlowPzELMAkGA1UEBhMCTk8xEzARBgNVBAgMClNvbWUtU3RhdGUxDTAL
    BgNVBAcMBE9zbG8xDDAKBgNVBAoMA1VpTzCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEAsnI3
    V1JKOl6Ghgzu5AOSsUFcJfefKP3y4lkVzW18T8AXkLKjNWQxJpW7h7kqeEYpcijAIV/xRMu8vEeb
    roadh9joOOD+ePL95VnAxcMxy3SdrGxRJ5RZwNf1l5UBxm416acIR78frlFX2kKz8kz0gmOmqFFe
    y3zZm7ZdiKw9fTsCAwEAATANBgkqhkiG9w0BAQUFAAOBgQBn6MlnleclAG/BVF4irljHag7OOopx
    TlQzb1V/mR0kYtsr6PZzCTJdwwU5YYWPG2hdqYRzyydlEGid5OPeCVtNQNQG8yKDCYeCS571jSbQ
    7Q0n0IR4mv7SeLq537mVBaaftV3HdnyoSLcXQpDep/akUOa8TpPwVg/+2U7TWNGJeQ==
</certificate>
```

with this:

``` xml
<public-key fingerprint="a54927a165a116f3d34cda8f84d6ab2d56c721a5c1e5351e049d3825aeec3962"/>
```

Extra benefit: reduced size of the registry reponse.

What do you think?

Also, it's worth to emphasize the fact that having invalid client certificates in the Manifest file won't affect the EWP Network as a whole - it will affect the **clients** (not servers) of a **single** EWP Host only. I guess it's not such a big usability loss and it would protect us from rogue CAs.

**Comment by andydowling-ie** [comment]
Good idea :-) Seems reasonable to me:
1. Strong binding on public keys
2. No need for a change in registry unless the institution's keys are re-generated
3. Smaller registry

As we can identify the association between the key fingerprint and the HEI (i.e. by tracking back up to the parent XML in the catalogue.xml), debugging shouldn't be too difficult.

**Comment by vgiralt** [comment]
The use of public key fingerprinting is an excellent idea. It was the use of X.509 certificates with all their expiration and CA added payload that has been causing grief in some SAML implementations. We can use trusted CA signed X.509 certificates as a transport for client certificates to enter the registry. It will also allow a transition to a DNSSec+DANE based distributed registry (or trust model) if we decide to move along that line without much hassle. Or, also, use DNSSec+DANE as a means for institutions to publish their keys to be automatically updated  in the registry if they have in-house know how, without impeding less technical participants to use other means to update the registry.

Great proposal!

**Comment by wrygiel** [comment]
Good to hear that you like it :) I'll try to update the docs next week.

**Comment by wrygiel** [comment]
One more gotcha: Should we allow or reject **expired client certificates** in this model?

You could say - it doesn't matter - and you would be right. But I just wonder which option would be less confusing for a newbie...

**Comment by wrygiel** [comment]
> It was the use of X.509 certificates with all their expiration and CA added payload that has been causing grief in some SAML implementations.

Could you expand on this? Were these problems caused by hosting a) server certificates or b) client certificates in their registry? If (a), then we won't have this problem in EWP anyway (we are thinking about hosting client certificates only, not the server ones).

**Comment by wrygiel** [comment]
> No need for a change in registry unless the institution's keys are re-generated

I did some research, and it would seem that most admins **do** change their keys every time they renew their certificate. If this is true, then replacing `<certificate>` with `<public-key>` won't change much.

I just spent a couple of hours adding keys to the documentation (uncommited yet). And only after that, I started to realize, that some aspects of this solution may be quite confusing (e.g. the expiry problem mentioned above). Perhaps it would be less confusing to stick with the original **option B**, as proposed by Andy (instead of **B+** I have proposed)?

**Comment by wrygiel** [comment]
@mikesteez quoted [this page](https://spaces.internet2.edu/display/InCFederation/X.509+Certificates+in+Metadata) in https://github.com/erasmus-without-paper/ewp-specs-api-registry/issues/1#issuecomment-185092727, and - while reading it - I have stumbled on this:

> Theoretically, if all the relevant software systems could accept a public key without a certificate wrapper, InCommon would only need to include the public key of each entity. As it is, the certificate is a convenient container for the public key, the critical element being that the key is bound to a particular entity in the metadata.

It would seem that a similar solution to our **option B+** has been considered there, but they have concluded that their equivalent of our **option B** would be the way to go.

**Comment by wrygiel** [comment]
Released in Discovery Manifest API v2.0.0.
